runtime/caml/misc.h

```diff
@@ -75,28 +75,32 @@ typedef size_t asize_t;
 CAMLdeprecated_typedef(addr, char *);
 #endif /* CAML_INTERNALS */
 
-/* Noreturn is preserved for compatibility reasons.
-   Instead of the legacy GCC/Clang-only
-     foo Noreturn;
-   you should prefer
-     CAMLnoreturn_start foo CAMLnoreturn_end;
-   which supports both GCC/Clang and MSVC.
+/* Noreturn, CAMLnoreturn_start and CAMLnoreturn_end are preserved
+   for compatibility reasons.  Instead, we recommend using the CAMLnoret
+   macro, to be added as a modifier at the beginning of the
+   function definition or declaration.
 
    Note: CAMLnoreturn is a different macro defined in memory.h,
-   to be used in function bodies rather than as a prototype attribute.
+   to be used in function bodies rather than as a function attribute.
 */
+#if __STDC_VERSION__ >= 202300L || __cplusplus >= 201103L
+  #define CAMLnoret [[noreturn]]
+#elif __STDC_VERSION__ >= 201112L
+  #define CAMLnoret _Noreturn
+#elif defined(__GNUC__)
+  #define CAMLnoret  __attribute__ ((noreturn))
```

Comment:
```suggestion
#elif defined(__GNUC__) /* Header processed in C++ mode */
  #define CAMLnoret  __attribute__ ((noreturn))
```
or words to that effect. I got puzzled reading it, and I could imagine it being accidentally hoovered in future!

Reply:
No, C++ is not the (main) reason for this special case.  Some not too old, not too recent versions of GCC support all of C11 but default to gnu99 mode, so `__STDC_VERSION__` corresponds to C99 and the macros above don't select `_Noreturn`.